File: scm-csharp/scheme/primitives/PrimitiveParentPid.cs
using System;
using System.IO;
using System.Runtime.InteropServices;

namespace scheme;

public class PrimitiveParentPid : Primitive
{
    public override string Name() => "parent-pid";

    public override string Info() =>
        "Syntax: (parent-pid)\n" +
        "Library: (scm system)\n" +
        "Description: Returns the OS process id of the parent of the current Scheme\n" +
        "  process, or #f if it cannot be determined. On Linux this reads /proc/self/stat;\n" +
        "  on other platforms #f may be returned.\n" +
        "Example:\n" +
        "  (parent-pid) => 12340";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 0, 0);
        if (!RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) return Value.F;
        long? ppid = ProcUtil.ReadPpid("self");
        return ppid.HasValue ? (object)(long)ppid.Value : Value.F;
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveParentPid.java
package scheme.primitives;
import scheme.*;

public class PrimitiveParentPid extends Primitive {
    @Override
    public String name() { return "parent-pid"; }

    @Override
    public String info() {
        return "Syntax: (parent-pid)\n" +
               "Library: (scm system)\n" +
               "Description: Returns the OS process id of the parent of the current\n" +
               "  Scheme process, or #f if it cannot be determined.\n" +
               "Example:\n" +
               "  (parent-pid) => 12340";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 0, 0);
        var parent = ProcessHandle.current().parent();
        if (parent.isPresent()) return parent.get().pid();
        return Value.F;
    }
}
